torch_geometric/nn/glob/equilibrium_aggregation.py

```diff
@@ -0,0 +1,183 @@
+        self.layers = torch.nn.ModuleList([
+            torch.nn.Sequential(
+                torch.nn.Linear(in_size, out_size),
+                # torch.nn.LayerNorm(out_size),
+                torch.nn.Tanh())
+            for in_size, out_size in zip(sizes[:-1], sizes[1:])
+        ])
+
+        self.res_trans = torch.nn.ModuleList([
+            torch.nn.Linear(in_channels, layer_size)
+            for layer_size in num_layers + [out_channels]
+        ])
+
+    def forward(self, x: torch.Tensor, y: torch.Tensor,
+                batch: Optional[torch.Tensor]) -> torch.Tensor:
+        if batch is None:
+            inp = torch.cat([x, y.expand(x.size(0), -1)], dim=1)
+        else:
+            inp = torch.cat([x, y[batch]], dim=1)
+
+        h = inp
+        for layer, res in zip(self.layers, self.res_trans):
+            h = layer(h)
+            h = res(inp) + h
+        return h.sum()
+
+
+class MomentumOptimizer(torch.nn.Module):
+    r"""
+    Provides an inner loop optimizer for the implicitly defined output
+    layer. It is based on an unrolled Nesterov momentum algorithm.
+
+    Args:
+        learning_rate (flaot): learning rate for optimizer.
+        momentum (float): momentum for optimizer.
+        learnable (bool): If :obj:`True` then the :obj:`learning_rate` and
+            :obj:`momentum` will be learnable parameters. If False they
+            are fixed. (default: :obj:`True`)
+    """
+    def __init__(self, learning_rate: float = 0.1, momentum: float = 0.9,
+                 learnable: bool = True):
+        super().__init__()
+
+        self._initial_lr = learning_rate
+        self._initial_mom = momentum
+        self._lr = torch.nn.Parameter(torch.Tensor([learning_rate]),
+                                      requires_grad=learnable)
+        self._mom = torch.nn.Parameter(torch.Tensor([momentum]),
+                                       requires_grad=learnable)
+        self.softplus = torch.nn.Softplus()
+        self.sigmoid = torch.nn.Sigmoid()
+        self.register_full_backward_hook(self.gradient_hook)
+
+    @staticmethod
+    def gradient_hook(module, grad_input, grad_output):
+        pass
+
+    def reset_parameters(self):
+        self._lr.data.fill_(self._initial_lr)
+        self._mom.data.fill_(self._initial_mom)
+
+    @property
+    def learning_rate(self):
+        return self.softplus(self._lr)
+
+    @property
+    def momentum(self):
+        return self.sigmoid(self._mom)
+
+    def forward(
+        self,
+        x: torch.Tensor,
+        y: torch.Tensor,
+        batch: Optional[torch.Tensor],
+        func: Callable[[torch.Tensor, torch.Tensor, Optional[torch.Tensor]],
+                       torch.Tensor],
+        iterations: int = 5,
+    ) -> Tuple[torch.Tensor, float]:
+
+        momentum = torch.zeros_like(y)
+        for _ in range(iterations):
+            val = func(x, y, batch)
+            grad = torch.autograd.grad(val, y, create_graph=True,
```

Comment:
The authors of the paper suggest adding an axillary loss on the `grad` here (`grad.square().sum()/iterations`) to be precise.